R/teal_data.R

```diff
@@ -36,27 +36,41 @@ teal_data <- function(...,
                       code = "",
                       check = FALSE) {
   data_objects <- list(...)
-  checkmate::assert_list(
-    data_objects,
-    types = c("TealDataset", "TealDatasetConnector", "TealDataConnector")
-  )
   if (inherits(join_keys, "JoinKeySet")) {
     join_keys <- teal.data::join_keys(join_keys)
   }
+  if (
+    checkmate::test_list(data_objects, types = c("TealDataConnector", "TealDataset", "TealDatasetConnector"))
+  ) {
+    warning("Using TealDatasetConnector and TealDataset is deprecated, please just include data directly.")
+    update_join_keys_to_primary(data_objects, join_keys)
 
-  update_join_keys_to_primary(data_objects, join_keys)
-
-  x <- TealData$new(..., check = check, join_keys = join_keys)
+    x <- TealData$new(..., check = check, join_keys = join_keys)
+    if (length(code) > 0 && !identical(code, "")) {
+      x$set_pull_code(code = code)
+    }
+    x$check_reproducibility()
+    x$check_metadata()
 
-  if (length(code) > 0 && !identical(code, "")) {
-    x$set_pull_code(code = code)
+    if (is_pulled(x)) {
+      new_tdata(
+        env = lapply(x$get_datasets(), function(x) x$get_raw_data()),
+        code = x$get_code(),
+        keys = x$get_join_keys()
+      )
+    } else {
+      x
+    }
+  } else {
+    new_tdata(
```

Comment:
What do you think about protecting against unnamed data parameters?

The error is not very user friendly

> names(x) must be a character vector of the same length as x

2 suggestions below if this makes sense:

### 1. outputs an error if arguments are not named


```suggestion
    new_tdata(
    if (!checkmate::test_names(names(data_objects), type = "named")) {
      stop("Dot (`...`) arguments on `teal_data()` must be named.")
    }
```

### 2. tries to extrapolate variable names and use it as name

```R
new_iris <- transform(iris, id = seq_len(nrow(iris)))
new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))

data = teal_data(
  new_iris,
  new_mtcars,
  code = "
      new_iris <- transform(iris, id = seq_len(nrow(iris)))
      new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))
    "
)

data = teal_data(
  new_iris = new_iris,
  new_mtcars,
  code = "
      new_iris <- transform(iris, id = seq_len(nrow(iris)))
      new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))
    "
)
```

```suggestion
    new_tdata(
    if (!checkmate::test_names(names(data_objects), type = "named")) {
      tryCatch({
        var_names <- vapply(substitute(list(...))[-1], deparse, character(1))
        unamed_ix <- names(data_objects) == ""
        names(data_objects)[unamed_ix] <- var_names[unamed_ix]
      }, error = function(err) stop("Dot (`...`) arguments on `teal_data()` must be named.")
      )
    }
```